nova_vm/src/ecmascript/builtins/structured_data/json_object.rs

```diff
@@ -68,3 +167,119 @@ impl JSONObject {
+
+                // 3. If newElement is undefined, then
+                if new_element.is_undefined() {
+                    // a. Perform ? val.[[Delete]](prop).
+                    val.internal_delete(agent, prop)?;
+                } else {
+                    // 4. Else,
+                    // a. Perform ? CreateDataProperty(val, prop, newElement).
+                    create_data_property(agent, val, prop, new_element)?;
+                }
+
+                // 5. Set I to I + 1.
+                i += 1;
+            }
+        } else {
+            // c. Else,
+            // i. Let keys be ? EnumerableOwnProperties(val, key).
+            let keys = enumerable_own_properties::<enumerable_properties_kind::Key>(agent, val)?;
+
+            // ii. For each String P of keys, do
+            for p in keys {
+                let p = PropertyKey::try_from(p).unwrap();
+                // 1. Let newElement be ? InternalizeJSONProperty(val, P, reviver).
+                let new_element = internalize_json_property(agent, val, p, reviver)?;
+
+                // 2. If newElement is undefined, then
+                if new_element.is_undefined() {
+                    // a. Perform ? val.[[Delete]](P).
+                    val.internal_delete(agent, p)?;
+                } else {
+                    // 3. Else,
+                    // a. Perform ? CreateDataProperty(val, P, newElement).
+                    create_data_property(agent, val, p, new_element)?;
+                }
+            }
+        }
+    }
+
+    // 3. Return ? Call(reviver, holder, « name, val »).
+    call_function(
+        agent,
+        reviver,
+        holder.into_value(),
+        Some(ArgumentsList(&[name.into_value(), val])),
+    )
+}
+
+pub(crate) fn value_from_json(agent: &mut Agent, json: &sonic_rs::Value) -> JsResult<Value> {
+    match json.get_type() {
+        sonic_rs::JsonType::Null => Ok(Value::Null),
+        sonic_rs::JsonType::Boolean => Ok(Value::Boolean(json.is_true())),
+        sonic_rs::JsonType::Number => Ok(Number::from_f64(agent, json.as_f64().unwrap()).into()),
+        sonic_rs::JsonType::String => Ok(String::from_str(agent, json.as_str().unwrap()).into()),
+        sonic_rs::JsonType::Array => {
+            let json_array = json.as_array().unwrap();
+            let len = json_array.len();
+            let array_obj = array_create(agent, len, len, None)?;
+            for (i, value) in json_array.iter().enumerate() {
+                let prop = PropertyKey::from(SmallInteger::try_from(i as i64).unwrap());
+                let js_value = value_from_json(agent, value)?;
+                create_data_property(agent, array_obj, prop, js_value)?;
+            }
+            Ok(array_obj.into())
+        }
+        sonic_rs::JsonType::Object => {
+            let json_object = json.as_object().unwrap();
+            let object =
+                ordinary_object_create_with_intrinsics(agent, Some(ProtoIntrinsics::Object), None);
```

Comment:
**thought:** We need some way to reserve the backing store size before we go into the loop. That way this won't eg. first fill a backing store of 8 property values, then copy them into a larger backing store and continue filling that.